src/main/java/io/getunleash/impactmetrics/CounterImpl.java

```diff
@@ -0,0 +1,75 @@
+package io.getunleash.impactmetrics;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+public class CounterImpl implements Counter {
+    private final String name;
+    private final String help;
+    private final ConcurrentHashMap<String, Long> values = new ConcurrentHashMap<>();
+
+    public CounterImpl(String name, String help) {
+        this.name = name;
+        this.help = help;
+    }
+
+    @Override
+    public void inc() {
+        inc(1, null);
+    }
+
+    @Override
+    public void inc(long value) {
+        inc(value, null);
+    }
+
+    @Override
+    public void inc(long value, Map<String, String> labels) {
+        String key = getLabelKey(labels);
+        values.compute(key, (k, current) -> (current == null ? 0L : current) + value);
+    }
+
+    public CollectedMetric collect() {
+        List<NumericMetricSample> samples = new ArrayList<>();
+
+        values.forEach(
+                (key, value) -> {
+                    samples.add(new NumericMetricSample(parseLabelKey(key), value));
+                });
+        values.clear();
+
+        if (samples.isEmpty()) {
+            samples.add(new NumericMetricSample(Collections.emptyMap(), 0L));
+        }
+
+        return new CollectedMetric(name, help, MetricType.COUNTER, samples);
+    }
+
+    static String getLabelKey(Map<String, String> labels) {
+        if (labels == null || labels.isEmpty()) {
+            return "";
+        }
+        return labels.entrySet().stream()
+                .sorted(Map.Entry.comparingByKey())
+                .map(e -> e.getKey() + "=" + e.getValue())
+                .collect(Collectors.joining(","));
+    }
+
+    static Map<String, String> parseLabelKey(String key) {
+        if (key == null || key.isEmpty()) {
+            return Collections.emptyMap();
+        }
+        ConcurrentHashMap<String, String> labels = new ConcurrentHashMap<>();
```

Comment:
you're not sharing this so Map<String, String> labels = new HashMap<>(); would suffice.